network/src/peer_manager/mod.rs

```diff
@@ -68,6 +69,11 @@ pub enum PeerManagerRequest {
     SendRpc(PeerId, #[serde(skip)] OutboundRpcRequest),
     /// Fire-and-forget style message send to a remote peer.
     SendDirectSend(PeerId, #[serde(skip)] Message),
+    /// Internal request to get supported protocols from the connection.
+    SupportedProtocols(
+        PeerId,
+        #[serde(skip)] oneshot::Sender<Result<SupportedProtocols, PeerManagerError>>,
+    ),
```

Comment:
I think we push the supported protocols up through the `NewPeer`/`LostPeer` events? The consensus network task can maintain a map `PeerId` -> `SupportedProtocols` or w/e and use that to decide what message encoding to send

Reply:
ah good to know